community: add token_limit_retriever

humblemat810 opened the pull request:
- Description: Added token_limit_retriever that limit the number of tokens retreived from retriever
- Issue: NIL, IDEAs #22941
- Dependencies: Test: [Optional] tiktoken
- Twitter handle: @ChanHungCheung

humblemat810 committed:
feat: token limiting retriever

token_limit_retriever limit the retrieved document token count

Useful when using token-count based cost models of api.

Also, even when no cost model involved, models still have token
limit and can reserve token for model to output.

humblemat810 committed:
test: improve test speed, refracter serial test to parallel

humblemat810 committed:
linting: run make lint

humblemat810 committed:
format: run ruff format

hwchase17 reviewed (commented):
reviewing internally. question is whether we want to support retriever classes like this, or just opt for adding a function and then instructing people to do `retriever | function`

humblemat810 commented:
1. the ensemble retriever existing is taking the same approach, wrapping around existing retrievers
2. the function approach will be a func that take a list of retriever outputs with each a list of retriever output.
consistency or not? or need to rewrite it into a function?

3. However, a functional approach has been written and test, either should work, but which one?
should push another pull request for the functional approach?

4. Should the functional form be put in langchain_community.callback? (utils also in callbacks, is it too generic?)
5. or make it a runnable lambda?

below is the functional form
token_limit.py
```python3
__all__ = ["token_limiter"]

from typing import Callable, List, Optional
from langchain.pydantic_v1 import validate_arguments
from langchain.schema.document import Document



from functools import partial
from copy import deepcopy
@validate_arguments
def token_limiter(token_limit, token_cutoff_strategy: str, 
                                     tokeniser_callback: Callable,
                                     remaining_token_fillup_callback: Optional[Callable[[str, int], str]] = None
                                     )-> Callable[[List[Document]], List[Document]]:
    if not callable(remaining_token_fillup_callback) ^ (
        token_cutoff_strategy == "complete_document"): 
        raise ValueError(
            'cannot set token_cutoff_strategy to be "complete_document" '
            "and set cutoff function at the same time."
            
        )
    if not( token_cutoff_strategy in [
            "complete_document",
            "partial_document"
        ]):
        raise(ValueError(f'token_cutoff_strategy must be "complete_document" or "partial_document"' ))


    @validate_arguments
    def wrapped(docs: List[Document], 
                tokeniser_callback: Callable,
                remaining_token_fillup_callback: Optional[Callable[[str, int], str]] = None
                ) -> List[Document]:
        # prevent late binding issues
        _tokeniser_callback = deepcopy(tokeniser_callback)
        _remaining_token_fillup_callback = deepcopy(remaining_token_fillup_callback)
        cummulative_token_len = 0
        len_doc_token = 0
        i = 0
        len_doc_token = len(_tokeniser_callback(docs[i].page_content))
        while cummulative_token_len + len_doc_token <= token_limit:
            cummulative_token_len += len_doc_token

            # check next
            i += 1
            if i == len(docs):
                break
            len_doc_token = len(_tokeniser_callback(docs[i].page_content))

        tokens_remaining = token_limit - cummulative_token_len
        to_return = docs[:i]
        if token_cutoff_strategy == "partial_document":
            if tokens_remaining > 0 and i < len(docs):
                # chop final document to fit the remaining token
                to_return.append(
                    Document(
                        _remaining_token_fillup_callback(
                            docs[i].page_content, tokens_remaining
                        )
                    )
                )

        return to_return
    
    return partial(wrapped, 
                   tokeniser_callback = tokeniser_callback, 
                   remaining_token_fillup_callback = remaining_token_fillup_callback)

```
and corresponding test:
```python3
import asyncio
from typing import Dict, List

import pytest
from langchain.pydantic_v1 import root_validator
from langchain_core.callbacks import CallbackManagerForRetrieverRun
from langchain_core.documents import Document
from langchain_core.retrievers import BaseRetriever
from token_limit import token_limiter


@pytest.fixture(scope="module")
def toy_retriever():
    class ToyRetriever(BaseRetriever):
        """A toy retriever that contains the top k documents
        that contain the user query.

        This retriever only implements the sync method _get_relevant_documents.

        If the retriever were to involve file access or network access, it could benefit
        from a native async implementation of `_aget_relevant_documents`.

        As usual, with Runnables, there's a default async implementation that's provided
        that delegates to the sync implementation running on another thread.
        """

        documents: List[Document] = [
            Document("this is a nice day"),
            Document("Peter looks great today"),
            Document("Will AI take over the world?"),
            Document("Langchain is a useful community contributed library."),
        ]

        """List of documents to retrieve from."""
        k: int
        """Number of top results to return"""

        @root_validator()
        def validate_environment(cls, values: Dict) -> Dict:
            return values

        def _get_relevant_documents(
            self, query: str, *, run_manager: CallbackManagerForRetrieverRun
        ) -> List[Document]:
            """Sync implementations for retriever."""
            matching_documents = []
            for document in self.documents:
                if len(matching_documents) > self.k:
                    return matching_documents

                if query.lower() in document.page_content.lower():
                    matching_documents.append(document)
            return matching_documents

    yield ToyRetriever(k=4)


@pytest.mark.requires("tiktoken")
@pytest.fixture(scope="module")
def get_tokenize_tiktoken():
    def tokenize_tiktoken(text):
        import tiktoken

        enc = tiktoken.encoding_for_model("gpt-3.5-turbo")
        tokens = enc.encode(text)
        return tokens

    yield tokenize_tiktoken


@pytest.mark.requires("tiktoken")
@pytest.fixture(scope="module")
def get_decode_tiktoken_upto():
    def decode_tiktoken_upto(text: str, n_token_left: int):
        import tiktoken

        enc = tiktoken.encoding_for_model("gpt-3.5-turbo")
        return enc.decode(enc.encode(text)[:n_token_left])

    yield decode_tiktoken_upto


@pytest.mark.requires("tiktoken")
def test_tiktoken_partial(
    toy_retriever, get_tokenize_tiktoken, get_decode_tiktoken_upto
):
    
    limiter_func = token_limiter(
        token_limit=20,
        remaining_token_fillup_callback=get_decode_tiktoken_upto,
        tokeniser_callback=get_tokenize_tiktoken,
        token_cutoff_strategy="partial_document",
    )
    simple_token_limit_chain = toy_retriever | limiter_func
    res = simple_token_limit_chain.invoke("a")
    assert len(res) == 4


def test_tiktoken_complete_parameter_error(
    toy_retriever, get_tokenize_tiktoken, get_decode_tiktoken_upto
):
    from pydantic.v1.error_wrappers import ValidationError
    # test for invalid optoins
    try:
        limiter_func = token_limiter(
            token_limit=20,
            remaining_token_fillup_callback=get_decode_tiktoken_upto,
            tokeniser_callback=get_tokenize_tiktoken,
            token_cutoff_strategy="complete_document"
        )
        simple_token_limit_chain = toy_retriever | limiter_func
        res = simple_token_limit_chain.invoke("a")
        assert len(res) == 3
    except ValueError as ve:
        assert str(ve) == ('cannot set token_cutoff_strategy to be "complete_document" '
                          "and set cutoff function at the same time.")
        \
    

def test_tiktoken_complete_parameter_correct(toy_retriever, get_tokenize_tiktoken):
    # test for complete document options
    limiter_func = token_limiter(
        token_limit=20,
        tokeniser_callback=get_tokenize_tiktoken,
        token_cutoff_strategy="complete_document",
    )
    simple_token_limit_chain = toy_retriever | limiter_func
    res = simple_token_limit_chain.invoke("a")
    assert len(res) == 3


def test_no_dependency(toy_retriever):

    def tokenize(text):
        return text.split(" ")

    def decode_token_upto(text: str, n_token_left: int):
        return " ".join(text.split(" ")[:n_token_left])

    limiter_func = token_limiter(
        token_limit=20,
        tokeniser_callback=tokenize,
        token_cutoff_strategy="complete_document",
    )
    simple_token_limit_chain = toy_retriever | limiter_func
    res = simple_token_limit_chain.invoke("a")
    assert len(res) == 3
    res = asyncio.run(simple_token_limit_chain.ainvoke("a"))
    assert len(res) == 3
    limiter_func = token_limiter(
        token_limit=20,
        remaining_token_fillup_callback=decode_token_upto,
        tokeniser_callback=tokenize,
        token_cutoff_strategy="partial_document",
    )
    simple_token_limit_chain = toy_retriever | limiter_func
    res = simple_token_limit_chain.invoke("a")
    assert len(res) == 4
    res = asyncio.run(simple_token_limit_chain.ainvoke("a"))
    assert len(res) == 4


def run_tests():
    pytest.main([__file__])


if __name__ == "__main__":
    run_tests()
```
